### PR TITLE
feat(merge): merge multiple branches in the same time

### LINKED ISSUE #4: UCP: merge multiple branches in the same time
## Description

The auto merge jobs should be scheduled in order, because the earlier merged PR will make an update branch patch to else. However, PRs base to different branches can be tested and merged in the same time.

## Score

- 200

## Mentor

@you06 

## Recommended Skills

- Go

## Learning Materials

- [go-github](https://github.com/google/go-github/)
- [merge provider code](https://github.com/pingcap-incubator/cherry-bot/tree/master/pkg/providers/merge)

### PR DESCRIPTION
**Pull request by Kuri-su**
UCP #4 

## What problem does this PR solve?
fix #4 

## What is changed and how it works?
I'm add a classifier before  'CheckPR()' function. It will classify the PRs in order according to the different 'PR Base' and then process them in parallel.

### CODE REVIEW

**Comment by Kuri-su** [comment]
@you06 PTAL

**Comment by you06** [comment]
@Kuri-su thanks for this PR, here I have some reviews.

The sha values of every PR are different, what you do in `classifyPR` actually parallel all PRs to be merged, you need to class them by ref, try `pr.GetBase().GetRef()`.

The merge job works by polling, check PR's CI status every 30 seconds. Suppose there are 3 PRs waiting for merging to `release-1.0` branch, we should always merge them in an specific order, it was promised by `model.Started`.

It's too costly if we class PRs by target branches via GitHub API. It's better to add a field like `target` in the database(we should know the target when receiving webhooks) to avoid frequently asking GitHub about this information.

**Comment by Kuri-su** [comment]
> @Kuri-su thanks for this PR, here I have some reviews.
> 
> The sha values of every PR are different, what you do in `classifyPR` actually parallel all PRs to be merged, you need to class them by ref, try `pr.GetBase().GetRef()`.
> 
> The merge job works by polling, check PR's CI status every 30 seconds. Suppose there are 3 PRs waiting for merging to `release-1.0` branch, we should always merge them in an specific order, it was promised by `model.Started`.
> 
> It's too costly if we class PRs by target branches via GitHub API. It's better to add a field like `target` in the database(we should know the target when receiving webhooks) to avoid frequently asking GitHub about this information.

OK, I got it

File: pkg/providers/merge/worker.go [hunk 1]
```diff
@@ -84,28 +85,58 @@ func (m *merge) startPolling() {
...
+			// for _, model := range jobs {
+			// 	if model.Started {
+			// 		job = model
+			// 	}
+			// }
+			// if job == nil {
+			// 	job = jobs[0]
+			// 	m.startJob(job)
+			// }
+
```
Comment (Kuri-su) [inline pkg/providers/merge/worker.go:98]: This code maybe was wrong?

the true like this?
```go
            var job *AutoMerge
            for _, model := range jobs {
                if model.Started {
                    job = model
                    break
                }
            }
            if job == nil {
                job = jobs[0]
                m.startJob(job)
            }
```

Comment (Kuri-su) [reply pkg/providers/merge/worker.go:98]: @you06 PTAL :smile:

**Comment by Kuri-su** [comment]
I'm update the logic, PTAL, thx :smile: @you06

File: pkg/providers/merge/worker.go (hunk 1 above)
Comment (you06) [reply pkg/providers/merge/worker.go]: Good catch!

Comment (you06) [reply pkg/providers/merge/worker.go]: Fortunately, I was copying a pointer here instead of take address, but it should be break!

**Comment by you06** [comment]
Mostly LGTM, I'll test this PR later.
And you may also update the SQL file.
https://github.com/pingcap-incubator/cherry-bot/blob/c8729a32b4feec75a1e1853f1e6d9648a2ff0e4b/bot.sql#L68-L81

**Comment by Kuri-su** [comment]
@you06 OK, I'm done

**Comment by you06** [comment]
@Kuri-su I've tested with your code, there are some problems.

The auto merge job can be trigger by labeled with a specific label like "can merge" either comment with "/merge", when use comment to trigger this job, you need to write the base ref into the database also, check out it at `merge/event.go:ProcessIssueCommentEvent`.

There will be a notice like "Your auto merge job has been accepted, waiting for 10, 11" when the current auto merge job need to wait for other jobs finished, now we need to notice for base ref the only instead of commenting with all pull requests, check out it at `merge/requests.go:queueComment`.

**Comment by Kuri-su** [comment]
@you06 OK, i'm done

**Comment by you06** [comment]
Awesome, it works! Thanks @Kuri-su !

**Pull request merged**